src/RegionHintRecommender.cpp

```diff
@@ -0,0 +1,128 @@
+            throw Exception("RegionHintRecommenderImp::" + std::string(__func__) +
+                            ": Frequency map file format is incorrect: " + err + ".",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        if (fmap_json.is_null() || !fmap_json.is_object()) {
+            throw Exception("RegionHintRecommenderImp::" + std::string(__func__) +
+                            ": Frequency map file format is incorrect: object expected.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        for (const auto &row : fmap_json.object_items()) {
+            if (!row.second.is_array() || row.second.array_items().size() == 0) {
+                throw Exception("RegionHintRecommenderImp::" + std::string(__func__) +
+                                ": Frequency map file format is incorrect: region keys "
+                                "must contain an array of numbers.",
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            m_freq_map[row.first].resize(row.second.array_items().size());
+            for (size_t idx = 0; idx < m_freq_map[row.first].size(); ++idx) {
+                if (!row.second[idx].is_number()) {
+                    throw Exception("RegionHintRecommenderImp::" + std::string(__func__) +
+                                    ": Non-numeric value found in frequencies for "
+                                    "region: \"" + row.first + "\".",
+                                    GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+                }
+                m_freq_map[row.first][idx] = row.second[idx].number_value();
+            }
+        }
+
+        if (m_freq_map.empty()) {
+            throw Exception("RegionHintRecommenderImp::" + std::string(__func__) +
+                            ": Frequency map file must contain a frequency map.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+    }
+
+    double RegionHintRecommenderImp::recommend_frequency(
+            const std::map<std::string, double> &nn_output,
+            double phi) const {
+        for (const auto &kv : nn_output ) {
+            if (std::isnan(kv.second)) {
+                return NAN;
+            }
+        }
+
+        double zz = 0;
+        double freq = 0;
+        for (const auto &[region_name, probability] : nn_output) {
+            size_t phi_idx = static_cast<size_t>(
+                                std::floor(phi * (m_freq_map.at(region_name).size() - 1)));
+            freq += exp(probability) * m_freq_map.at(region_name).at(phi_idx);
+            zz += exp(probability);
+        }
+        freq = freq / zz;
+
+        if (std::isnan(freq)) {
+            freq = m_max_freq;
+        }
+
+        if (freq > m_max_freq) {
+            freq = m_max_freq;
+        }
+        if (freq < m_min_freq) {
+            freq = m_min_freq;
+        }
+
+        return freq * 1e8;
```

Comment:
I expected to see 1e9 here.  What units is freq in before the conversion?  It must be MHz for this to make sense.
The intention is to have the output in Hz?  Is this correct?
It would be nice to see whatever we multiply by here either explained by a nearby comment or captured as a constexpr or similar instead of a magic number that only shows up when we go to return.

Reply:
Confirmed that there is no numerical reason to divert from SI units here. We're going forward with using Hz everywhere. Code has been changed to reflect that.